apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/DateTimestampReadTests.cs

```diff
@@ -0,0 +1,99 @@
+                    DateTime.SpecifyKind(new DateTime(2020, 11, 29), DateTimeKind.Local),
+                    new DateTime(2020, 12, 29, 0, 0, 0, DateTimeKind.Utc),
+                    new DateTime(2020, 12, 29, 0, 0, 0, DateTimeKind.Local),
+                    DateTime.Today,
+                })
+                {
+                    // Write the date value and read it back.
+                    await connection.CreateInsertOrUpdateCommand(_fixture.TableName,
+                        new SpannerParameterCollection
+                        {
+                            new SpannerParameter("DateValue", SpannerDbType.Date, date),
+                        }
+                    ).ExecuteNonQueryAsync();
+                    using (var reader = await connection.CreateSelectCommand($"SELECT DateValue FROM {_fixture.TableName}").ExecuteReaderAsync())
+                    {
+                        Assert.True(await reader.ReadAsync());
+                        Assert.Equal(date, reader.GetFieldValue<DateTime>("DateValue"));
+                        Assert.False(await reader.ReadAsync());
+                    }
+                }
+            }
+        }
+
+        [Fact]
+        public async void WriteTimestampThenRead_ShouldBeEqual()
+        {
+            using (var connection = _fixture.GetConnection())
+            {
+                foreach (var timestamp in new DateTime[] {
+                    new DateTime(2020, 9, 29, 18, 51, 10),
+                    DateTime.SpecifyKind(new DateTime(2020, 10, 29, 18, 51, 10), DateTimeKind.Utc),
+                    DateTime.SpecifyKind(new DateTime(2020, 11, 29, 18, 51, 10), DateTimeKind.Local),
+                    new DateTime(2020, 10, 29, 18, 51, 10, DateTimeKind.Utc),
+                    new DateTime(2020, 11, 29, 18, 51, 10, DateTimeKind.Local),
+                    DateTime.Now,
+                    DateTime.UtcNow,
+                })
+                {
+                    // Write the timestamp value and read it back.
+                    await connection.CreateInsertOrUpdateCommand(_fixture.TableName,
+                        new SpannerParameterCollection
+                        {
+                            new SpannerParameter("TimestampValue", SpannerDbType.Timestamp, timestamp),
+                        }
+                    ).ExecuteNonQueryAsync();
+                    using (var reader = await connection.CreateSelectCommand($"SELECT TimestampValue FROM {_fixture.TableName}").ExecuteReaderAsync())
+                    {
+                        Assert.True(await reader.ReadAsync());
+                        var dbTimestamp = reader.GetFieldValue<DateTime>("TimestampValue");
+                        // Cloud Spanner always stores the timestamp in UTC, so if the original input value was
+                        // a local time, we need to convert the value that we read in the database back to local time.
+                        if (timestamp.Kind == DateTimeKind.Local)
```

Comment:
Better if there are no iffs, just convert the timestamp to UTC always before the comparison.

Reply:
Actually, that does not work (assuming that you mean calling `timestamp.ToUniversalTime()` on the expected timestamp). The reason is that:
1. If I write a `DateTime` instance with `DateTimeKind.Unspecified` to the database, that is interpreted by the client library as a `DateTime` that is already in UTC. So writing `new DateTime(2020, 9, 29, 18, 51, 10)` to the database inserts `20200929T18:51:10Z`.
2. Calling `DateTime.ToUniversalTime()` on a `DateTime` instance with `DateTimeKind.Unspecified` will make .NET treat the `DateTime` as if it was in the local timestamp.

So there is a discrepancy between how the Cloud Spanner client library interprets an unspecified `DateTime` and how .NET interprets such as a `DateTime`. I think the current behavior of the Spanner client library is right in interpreting an unspecified `DateTime` as already being in UTC, as Cloud Spanner makes it clear that `TIMESTAMP` values are stored in UTC.